File: Java/app/src/main/java/com/tyhoo/mvvm/java/util/ViewModelFactory.java
package com.tyhoo.mvvm.java.util;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.tyhoo.mvvm.java.ui.MainViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private static volatile ViewModelFactory INSTANCE;

    public static ViewModelFactory getInstance() {
        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory();
                }
            }
        }
        return INSTANCE;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new MainViewModel();
    }
}

File: Kotlin/app/src/main/java/com/tyhoo/mvvm/kotlin/util/ViewModelFactory.kt
package com.tyhoo.mvvm.kotlin.util

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.tyhoo.mvvm.kotlin.ui.MainViewModel

class ViewModelFactory : ViewModelProvider.NewInstanceFactory() {

    override fun <T : ViewModel> create(modelClass: Class<T>) = MainViewModel() as T

    companion object {
        @Volatile
        private var INSTANCE: ViewModelFactory? = null

        fun getInstance() = INSTANCE ?: synchronized(ViewModelFactory::class.java) {
            INSTANCE ?: ViewModelFactory().also { INSTANCE = it }
        }
    }
}